torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,474 @@
+                ReanalyzeWorker(
+                    manager.Queue(2),
+                    reanalyze_queue,
+                    reanalyze_update_queue,
+                    model_id,
+                    config.training.reanalyze_device,
+                    reward_distribution,
+                    value_distribution,
+                    config.training.threads_per_worker,
+                )
+                for i in range(num_reanalyze_workers)
+            ]
+
+            trials_completed = 0
+            running_stats = RunningStats()
+
+            xp_tracker.log_params(
+                config.environment,
+                config.actor,
+                config.training,
+            )
+
+            trial_configs = make_trial_configs(run_session, config, model_id, -1)
+
+            total_samples = 0
+            training_step = 0
+            run_total_reward = 0
+            epoch_idx = 0
+
+            workers = [train_worker, replay_buffer] + reanalyze_workers
+
+            for worker in reanalyze_workers:
+                worker.update_agent(agent)
+
+            try:
+                for worker in workers:
+                    worker.start()
+
+                start_time = time.time()
+
+                sample_generator = run_session.start_trials_and_wait_for_termination(
+                    trial_configs,
+                    max_parallel_trials=config.training.max_parallel_trials,
+                )
+
+                async for _step, timestamp, trial_id, _tick, (sample, total_reward) in sample_generator:
+                    replay_buffer.add_sample(trial_id, sample)
+                    total_samples += 1
+
+                    for worker in workers:
+                        assert worker.is_alive()
+
+                    if sample.done:
+                        trials_completed += 1
+                        xp_tracker.log_metrics(
+                            timestamp, total_samples, trial_total_reward=total_reward, trials_completed=trials_completed
+                        )
+                        run_total_reward += total_reward
+
+                    if replay_buffer.size() <= config.training.min_replay_buffer_size:
+                        continue
+
+                    while not results_queue.empty():
+                        try:
+                            info, serialized_model = results_queue.get_nowait()
+                            if serialized_model is not None:
+                                epoch_idx += 1
+                                agent = MuZeroAgent.load(io.BytesIO(serialized_model), "cpu")
+                                version_info = await self.publish_version(
+                                    model_id,
+                                    agent,
+                                    epoch_idx=epoch_idx,
+                                    total_samples=total_samples,
+                                    environment_specs=config.environment.specs,
+                                )
```

Comment:
It seems like we could something better that serialize -> put in queue -> deserialize -> publish -> serialize -> send to model registry.
Let's think about it once this is merged

Reply:
The seemingly unnecessary serialization is a workaround for very weird torch multiprocessing behaviour with `torch.nn.Module` objects. This was a last resort, I really tried everything that I could think of. We might want to add an option to the model registry client to publish directly the serialized version.